File: antia_numerical_methods/c/fun/gammap.c
/*	To calculate the incomplete Gamma function P(a,x) for
	positive real arguments
	It returns a value of -1 for a.le.0 or x<0

	a : (input) Argument for the complete Gamma function
	x : (input) Upper limit of integration defining the incomplete
			Gamma function

	Required routines : gamma, gammaln
*/
 

#include <math.h>

double gamma(double xg);
double gammaln(double xg);

double gammap(double a, double x)
{
	double c[500],b[500],f,s1,s2,g1,t,c1,c2;
	int i;

	if(a<=0 || x<0) return -1.0; 

	if(x<3)
	{
/*	Use power series */
        	s1=1; t=a; i=0;
		while(i<500 && fabs(t/(a+i))>1.e-14)
		{
			i=i+1; t=-t*x/i;
			s1=s1+t/(a+i);
		}
		if(a<140) f=pow(x,a)*s1/gamma(a+1);
		else f=0.0;
	}
	else if(a<1.2*x)
	{
/*	Use continued fraction */
        	c[0]=1; b[0]=x; c[1]=1; b[1]=x+1-a;
		c2=c[1]/b[1]; c1=0.0; i=0;
		while(i<500 && fabs(c1-c2)>1.e-12)
		{
			i=i+2; c1=c2;
			c[i]=x*c[i-1]+(i/2)*c[i-2];
			b[i]=x*b[i-1]+(i/2)*b[i-2];
			c[i+1]=c[i]+(i/2+1-a)*c[i-1];
			b[i+1]=b[i]+(i/2+1-a)*b[i-1];
			if(fabs(b[i+1])>1.e200)
			{
				c[i]=c[i]/1.e200; c[i+1]=c[i+1]/1.e200;
				b[i]=b[i]/1.e200; b[i+1]=b[i+1]/1.e200;
			}
			if(fabs(b[i+1])<1.e-200)
			{
				c[i]=c[i]*1.e200; c[i+1]=c[i+1]*1.e200;
				b[i]=b[i]*1.e200; b[i+1]=b[i+1]*1.e200;
			}
			c2=c[i+1]/b[i+1];
		}
		g1=-x+a*log(x)+log(c2)-gammaln(a); f=1-exp(g1);
	}
	else
	{
/*	Use the power series for a>x */
		s2=1; t=1; i=0;
		while(i<500 && fabs(t)>1.e-14)
		{
			i=i+1; t=t*x/(a+i); s2=s2+t;
		}
		g1=-x+a*log(x)+log(s2)-gammaln(a+1); f=exp(g1);
	}
	return f;
}

File: antia_numerical_methods@Beliavsky/f77/sub8/gammap.f
!	To calculate the incomplete Gamma function P(a,x) for
!	positive real arguments
!	It returns a value of -1 for a.le.0 or x<0
!
!	A : (input) Argument for the complete Gamma function
!	X : (input) Upper limit of integration defining the incomplete
!                       Gamma function
!
!	Required routines : GAMMA, GAMMAL

      FUNCTION GAMMAP(A,X)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION C(500),B(500)

      IF(A.LE.0.OR.X.LT.0) THEN
        GAMMAP=-1
        RETURN
      ENDIF

      IF(X.LT.3) THEN
!	Use power series
        S1=1
        T=A
        DO I=1,500
          T=-T*X/I
          S1=S1+T/(A+I)
          IF(ABS(T/(A+I)).LT.1.D-14) EXIT
        ENDDO
        IF(A.LT.140) THEN
          GAMMAP=X**A*S1/GAMMA(A+1)
        ELSE
          GAMMAP=0.0
        ENDIF
      ELSE IF(A.LT.1.2D0*X) THEN
!	Use continued fraction
        C(1)=1
        B(1)=X
        C(2)=1
        B(2)=X+1-A
        C1=C(2)/B(2)
        DO I=2,498,2
          C(I+1)=X*C(I)+(I/2)*C(I-1)
          B(I+1)=X*B(I)+(I/2)*B(I-1)
          C(I+2)=C(I+1)+(I/2+1-A)*C(I)
          B(I+2)=B(I+1)+(I/2+1-A)*B(I)
          IF(ABS(B(I+2)).GT.1.D200) THEN
            C(I+1)=C(I+1)/1.D200
            B(I+1)=B(I+1)/1.D200
            C(I+2)=C(I+2)/1.D200
            B(I+2)=B(I+2)/1.D200
          ENDIF
          IF(ABS(B(I+2)).LT.1.D-200) THEN
            C(I+1)=C(I+1)*1.D200
            B(I+1)=B(I+1)*1.D200
            C(I+2)=C(I+2)*1.D200
            B(I+2)=B(I+2)*1.D200
          ENDIF
          C2=C(I+2)/B(I+2)
          IF(ABS(C2-C1).LT.1.D-12) EXIT
          C1=C2
        ENDDO
        G1=-X+A*LOG(X)+LOG(C2)-GAMMAL(A)
        GAMMAP=1-EXP(G1)
      ELSE
!	Use the power series for a>x
        S2=1
        T=1
        DO I=1,500
          T=T*X/(A+I)
          S2=S2+T
          IF(ABS(T).LT.1.D-14) EXIT
        ENDDO
        G1=-X+A*LOG(X)+LOG(S2)-GAMMAL(A+1)
        GAMMAP=EXP(G1)
      ENDIF
      END
